File: scala/src/main/scala/com/github/copilot/CopilotSession.scala
// Copyright (c) Microsoft Corporation. All rights reserved.

package com.github.copilot

import io.circe.*
import io.circe.parser.*
import io.circe.syntax.*

import java.util.concurrent.ConcurrentHashMap
import scala.collection.mutable
import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.concurrent.duration.*
import scala.jdk.CollectionConverters.*
import scala.util.{Failure, Success, Try}

/**
 * Represents a single conversation session with the Copilot CLI.
 *
 * A session maintains conversation state, handles events, and manages tool execution.
 * Sessions are created via [[CopilotClient.createSession]] or resumed via
 * [[CopilotClient.resumeSession]].
 *
 * == Usage ==
 * {{{
 * val session = Await.result(client.createSession(), 30.seconds)
 *
 * // Subscribe to events
 * val unsub = session.on { event =>
 *   if event.`type` == SessionEventType.AssistantMessage then
 *     println(s"Assistant: ${event.data}")
 * }
 *
 * // Send and wait
 * val response = Await.result(
 *   session.sendAndWait(MessageOptions(prompt = "Hello!")),
 *   60.seconds
 * )
 *
 * unsub() // unsubscribe
 * Await.result(session.destroy(), 10.seconds)
 * }}}
 *
 * @param sessionId     The unique session identifier
 * @param connection    The JSON-RPC client for server communication
 * @param workspacePath Path to the workspace directory (when infinite sessions enabled)
 */
class CopilotSession(
  val sessionId: String,
  private val connection: JsonRpcClient,
  val workspacePath: Option[String] = None
)(using ec: ExecutionContext):

  // Event handlers
  private val eventHandlers = java.util.Collections.newSetFromMap(
    new ConcurrentHashMap[SessionEvent => Unit, java.lang.Boolean]()
  )
  private val typedEventHandlers = new ConcurrentHashMap[String, mutable.Set[SessionEvent => Unit]]()

  // Tool handlers
  private val toolHandlers = new ConcurrentHashMap[String, ToolHandler]()

  // Permission, user input, and hooks handlers
  @volatile private var permissionHandler: Option[PermissionHandler] = None
  @volatile private var userInputHandler: Option[UserInputHandler] = None
  @volatile private var hooks: Option[SessionHooks] = None

  // -------------------------------------------------------------------------
  // Sending Messages
  // -------------------------------------------------------------------------

  /**
   * Sends a message to the session.
   *
   * The message is enqueued for processing. Events are dispatched to registered handlers
   * as the assistant processes the message.
   *
   * @param options The message options
   * @return A Future with the message ID
   */
  def send(options: MessageOptions): Future[String] =
    val params = Json.obj(
      "sessionId" -> sessionId.asJson,
      "prompt" -> options.prompt.asJson,
      "attachments" -> options.attachments.asJson,
      "mode" -> options.mode.asJson,
      "responseFormat" -> options.responseFormat.asJson,
      "imageOptions" -> options.imageOptions.asJson,
    ).dropNullValues

    connection.sendRequest("session.send", params).map { result =>
      result.hcursor.get[String]("messageId").getOrElse(
        throw new RuntimeException("session.send response missing messageId")
      )
    }

  /**
   * Sends a message and waits until the session becomes idle.
   *
   * This is a convenience method that combines [[send]] with waiting for the
   * `session.idle` event. Events are still delivered to registered handlers while waiting.
   *
   * @param options The message options
   * @param timeout Maximum time to wait (default: 60 seconds)
   * @return A Future with the last assistant.message event, or None if no assistant message was received
   */
  def sendAndWait(
    options: MessageOptions,
    timeout: FiniteDuration = 60.seconds
  ): Future[Option[SessionEvent]] =
    val idlePromise = Promise[Option[SessionEvent]]()
    @volatile var lastAssistantMessage: Option[SessionEvent] = None

    // Register event handler BEFORE calling send to avoid race condition
    val unsubscribe = on { event =>
      event.`type` match
        case SessionEventType.AssistantMessage =>
          lastAssistantMessage = Some(event)
        case SessionEventType.SessionIdle =>
          idlePromise.trySuccess(lastAssistantMessage)
        case SessionEventType.SessionError =>
          val errorMsg = event.data.hcursor.get[String]("message").getOrElse("Unknown error")
          val errorStack = event.data.hcursor.get[String]("stack").toOption
          val ex = new RuntimeException(errorMsg)
          errorStack.foreach(s => ex.setStackTrace(Array.empty)) // stack from server
          idlePromise.tryFailure(ex)
        case _ => // ignore other events
    }

    // Create timeout Future
    val timeoutFuture = Future:
      Thread.sleep(timeout.toMillis)
      throw new java.util.concurrent.TimeoutException(
        s"Timeout after ${timeout.toMillis}ms waiting for session.idle"
      )

    val resultFuture = send(options).flatMap { _ =>
      Future.firstCompletedOf(Seq(idlePromise.future, timeoutFuture))
    }

    // Clean up on completion
    resultFuture.andThen { case _ => unsubscribe() }

  // -------------------------------------------------------------------------
  // Event Subscriptions
  // -------------------------------------------------------------------------

  /**
   * Subscribes to all events from this session.
   *
   * @param handler Callback for all session events
   * @return An unsubscribe function
   */
  def on(handler: SessionEvent => Unit): () => Unit =
    eventHandlers.add(handler)
    () => eventHandlers.remove(handler)

  /**
   * Subscribes to a specific event type.
   *
   * @param eventType The event type string (e.g., SessionEventType.AssistantMessage)
   * @param handler   Callback for matching events
   * @return An unsubscribe function
   */
  def on(eventType: String, handler: SessionEvent => Unit): () => Unit =
    typedEventHandlers
      .computeIfAbsent(eventType, _ => mutable.Set.empty)
      .addOne(handler)
    () =>
      val handlers = typedEventHandlers.get(eventType)
      if handlers != null then handlers.remove(handler)

  // -------------------------------------------------------------------------
  // Message History
  // -------------------------------------------------------------------------

  /**
   * Retrieves all events and messages from this session's history.
   *
   * @return A Future with all session events
   */
  def getMessages(): Future[List[SessionEvent]] =
    connection.sendRequest("session.getMessages", Json.obj("sessionId" -> sessionId.asJson)).flatMap { result =>
      result.hcursor.get[List[SessionEvent]]("events") match
        case Right(events) => Future.successful(events)
        case Left(err) => Future.failed(new RuntimeException(s"Failed to decode messages: $err"))
    }

  /**
   * Retrieves metadata for this session.
   *
   * @return A Future with the session metadata
   */
  def getMetadata(): Future[Json] =
    connection.sendRequest("session.getMetadata", Json.obj("sessionId" -> sessionId.asJson))

  // -------------------------------------------------------------------------
  // Session Lifecycle
  // -------------------------------------------------------------------------

  /**
   * Destroys this session and releases all associated resources.
   *
   * After calling this, the session can no longer be used. To continue the conversation,
   * use [[CopilotClient.resumeSession]] with the session ID.
   */
  def destroy(): Future[Unit] =
    connection.sendRequest("session.destroy", Json.obj("sessionId" -> sessionId.asJson)).map { _ =>
      eventHandlers.clear()
      typedEventHandlers.clear()
      toolHandlers.clear()
      permissionHandler = None
      userInputHandler = None
      hooks = None
    }

  /**
   * Aborts the currently processing message.
   *
   * The session remains valid and can be used for new messages.
   */
  def abort(): Future[Unit] =
    connection.sendRequest("session.abort", Json.obj("sessionId" -> sessionId.asJson)).map(_ => ())

  // -------------------------------------------------------------------------
  // Tool Registration (internal)
  // -------------------------------------------------------------------------

  /** Registers tool handlers. Called internally by CopilotClient. */
  private[copilot] def registerTools(tools: List[Tool]): Unit =
    toolHandlers.clear()
    tools.foreach(t => toolHandlers.put(t.name, t.handler))

  /** Gets a registered tool handler by name. */
  private[copilot] def getToolHandler(name: String): Option[ToolHandler] =
    Option(toolHandlers.get(name))

  /** Registers a permission handler. */
  private[copilot] def registerPermissionHandler(handler: PermissionHandler): Unit =
    permissionHandler = Some(handler)

  /** Registers a user input handler. */
  private[copilot] def registerUserInputHandler(handler: UserInputHandler): Unit =
    userInputHandler = Some(handler)

  /** Registers session hooks. */
  private[copilot] def registerHooks(h: SessionHooks): Unit =
    hooks = Some(h)

  // -------------------------------------------------------------------------
  // Event Dispatch (internal)
  // -------------------------------------------------------------------------

  /** Dispatches an event to all registered handlers. Called by CopilotClient. */
  private[copilot] def dispatchEvent(event: SessionEvent): Unit =
    // Typed handlers
    val typedSet = typedEventHandlers.get(event.`type`)
    if typedSet != null then
      typedSet.foreach { handler =>
        Try(handler(event)).failed.foreach(_ => ()) // ignore handler errors
      }
    // Wildcard handlers
    eventHandlers.forEach { handler =>
      Try(handler(event)).failed.foreach(_ => ()) // ignore handler errors
    }

  // -------------------------------------------------------------------------
  // Request Handlers (internal, called by CopilotClient)
  // -------------------------------------------------------------------------

  /** Handles a permission request from the server. */
  private[copilot] def handlePermissionRequest(requestJson: Json): Future[PermissionRequestResult] =
    permissionHandler match
      case None =>
        Future.successful(PermissionRequestResult(
          kind = PermissionRequestResult.DeniedNoApprovalRule
        ))
      case Some(handler) =>
        requestJson.as[PermissionRequest] match
          case Left(_) =>
            Future.successful(PermissionRequestResult(
              kind = PermissionRequestResult.DeniedNoApprovalRule
            ))
          case Right(request) =>
            handler(request, PermissionInvocation(sessionId)).recover {
              case _ => PermissionRequestResult(
                kind = PermissionRequestResult.DeniedNoApprovalRule
              )
            }

  /** Handles a user input request from the server. */
  private[copilot] def handleUserInputRequest(request: UserInputRequest): Future[Json] =
    userInputHandler match
      case None =>
        Future.failed(new RuntimeException("User input requested but no handler registered"))
      case Some(handler) =>
        handler(request, UserInputInvocation(sessionId)).map(_.asJson)

  /** Handles a hooks invocation from the server. */
  private[copilot] def handleHooksInvoke(hookType: String, input: Json): Future[Option[Json]] =
    hooks match
      case None => Future.successful(None)
      case Some(h) =>
        val invocation = HookInvocation(sessionId)
        hookType match
          case "preToolUse" =>
            h.onPreToolUse match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[PreToolUseHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case "postToolUse" =>
            h.onPostToolUse match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[PostToolUseHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case "userPromptSubmitted" =>
            h.onUserPromptSubmitted match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[UserPromptSubmittedHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case "sessionStart" =>
            h.onSessionStart match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[SessionStartHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case "sessionEnd" =>
            h.onSessionEnd match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[SessionEndHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case "errorOccurred" =>
            h.onErrorOccurred match
              case None => Future.successful(None)
              case Some(handler) =>
                input.as[ErrorOccurredHookInput] match
                  case Left(_) => Future.successful(None)
                  case Right(parsed) =>
                    handler(parsed, invocation).map(_.map(_.asJson)).recover { case _ => None }

          case _ => Future.successful(None)

File: swift/Sources/CopilotSDK/CopilotSession.swift
// ----------------------------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation. All rights reserved.
// ----------------------------------------------------------------------------------------------------

import Foundation

/// Represents a single conversation session with the Copilot CLI.
///
/// A session maintains conversation state, handles events, and manages tool execution.
/// Sessions are created via `CopilotClient.createSession()` or resumed via
/// `CopilotClient.resumeSession()`.
///
/// Example:
/// ```swift
/// let session = try await client.createSession()
///
/// session.on { event in
///     if event.type == "assistant.message" {
///         print(event.data["content"] as? String ?? "")
///     }
/// }
///
/// let response = try await session.sendAndWait(MessageOptions(prompt: "Hello!"))
/// try await session.destroy()
/// ```
public actor CopilotSession {
    // MARK: - Properties

    /// The unique identifier for this session.
    public let sessionId: String

    /// Path to the session workspace directory when infinite sessions are enabled.
    public let workspacePath: String?

    private let rpcClient: JsonRpcClient
    private var eventHandlers: [UUID: SessionEventHandler] = [:]
    private var typedEventHandlers: [String: [UUID: SessionEventHandler]] = [:]
    private var toolHandlers: [String: ToolHandler] = [:]
    private var permissionHandler: PermissionHandler?
    private var userInputHandler: UserInputHandler?
    private var hooks: SessionHooks?

    // MARK: - Init

    /// Creates a new CopilotSession.
    /// - Parameters:
    ///   - sessionId: The unique session identifier.
    ///   - rpcClient: The JSON-RPC client for communication.
    ///   - workspacePath: Optional workspace path for infinite sessions.
    init(sessionId: String, rpcClient: JsonRpcClient, workspacePath: String? = nil) {
        self.sessionId = sessionId
        self.rpcClient = rpcClient
        self.workspacePath = workspacePath
    }

    // MARK: - Send Message

    /// Sends a message to this session.
    ///
    /// - Parameter options: The message options including prompt and optional attachments.
    /// - Returns: The message ID of the sent message.
    @discardableResult
    public func send(_ options: MessageOptions) async throws -> String {
        var params: [String: Any] = [
            "sessionId": sessionId,
            "prompt": options.prompt,
        ]

        if let attachments = options.attachments {
            params["attachments"] = attachments.map { $0.toDict() }
        }
        if let mode = options.mode {
            params["mode"] = mode
        }
        if let responseFormat = options.responseFormat {
            params["responseFormat"] = responseFormat.rawValue
        }
        if let imageOptions = options.imageOptions {
            var imageDict: [String: Any] = [:]
            if let size = imageOptions.size { imageDict["size"] = size }
            if let quality = imageOptions.quality { imageDict["quality"] = quality }
            if let style = imageOptions.style { imageDict["style"] = style }
            params["imageOptions"] = imageDict
        }

        let response = try await rpcClient.sendRequest(method: "session.send", params: params)
        guard let messageId = response["messageId"] as? String else {
            throw CopilotError.invalidResponse("Missing messageId in session.send response")
        }
        return messageId
    }

    /// Sends a message and waits until the session becomes idle.
    ///
    /// This is a convenience method that combines `send` with waiting for the
    /// `session.idle` event. Events are still delivered to handlers while waiting.
    ///
    /// - Parameters:
    ///   - options: The message options.
    ///   - timeout: Timeout in seconds. Defaults to 60 seconds.
    /// - Returns: The final assistant message event, or nil if none was received.
    public func sendAndWait(
        _ options: MessageOptions,
        timeout: TimeInterval = 60
    ) async throws -> SessionEvent? {
        // Use a continuation-based approach: register event handler, send message,
        // wait for session.idle or session.error.

        return try await withCheckedThrowingContinuation { continuation in
            var lastAssistantMessage: SessionEvent?
            var resumed = false
            let handlerId = UUID()

            // Set up a timeout task
            let timeoutTask = Task {
                try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                if !resumed {
                    resumed = true
                    // Remove the handler
                    Task { await self.removeEventHandler(handlerId) }
                    continuation.resume(
                        throwing: CopilotError.timeout(
                            "Timeout after \(timeout)s waiting for session.idle"))
                }
            }

            // Register event handler BEFORE sending to avoid race conditions
            let handler: SessionEventHandler = { event in
                if event.type == "assistant.message" {
                    lastAssistantMessage = event
                } else if event.type == "session.idle" {
                    if !resumed {
                        resumed = true
                        timeoutTask.cancel()
                        Task { await self.removeEventHandler(handlerId) }
                        continuation.resume(returning: lastAssistantMessage)
                    }
                } else if event.type == "session.error" {
                    if !resumed {
                        resumed = true
                        timeoutTask.cancel()
                        Task { await self.removeEventHandler(handlerId) }
                        let msg = event.data["message"] as? String ?? "Unknown session error"
                        continuation.resume(throwing: CopilotError.operationFailed(msg))
                    }
                }
            }

            eventHandlers[handlerId] = handler

            // Now send the message
            Task {
                do {
                    try await self.send(options)
                } catch {
                    if !resumed {
                        resumed = true
                        timeoutTask.cancel()
                        await self.removeEventHandler(handlerId)
                        continuation.resume(throwing: error)
                    }
                }
            }
        }
    }

    // MARK: - Event Subscription

    /// Subscribes to all events from this session.
    ///
    /// - Parameter handler: A callback function that receives all session events.
    /// - Returns: An ID that can be used to unsubscribe via `removeEventHandler`.
    @discardableResult
    public func on(_ handler: @escaping SessionEventHandler) -> UUID {
        let id = UUID()
        eventHandlers[id] = handler
        return id
    }

    /// Subscribes to a specific event type from this session.
    ///
    /// - Parameters:
    ///   - eventType: The event type string (e.g., "assistant.message", "session.idle").
    ///   - handler: A callback function that receives events of that type.
    /// - Returns: An ID that can be used to unsubscribe.
    @discardableResult
    public func on(_ eventType: String, handler: @escaping SessionEventHandler) -> UUID {
        let id = UUID()
        if typedEventHandlers[eventType] == nil {
            typedEventHandlers[eventType] = [:]
        }
        typedEventHandlers[eventType]![id] = handler
        return id
    }

    /// Removes an event handler by its ID.
    public func removeEventHandler(_ id: UUID) {
        eventHandlers.removeValue(forKey: id)
        for key in typedEventHandlers.keys {
            typedEventHandlers[key]?.removeValue(forKey: id)
        }
    }

    // MARK: - Event Dispatch (Internal)

    /// Dispatches an event to all registered handlers.
    func dispatchEvent(_ event: SessionEvent) {
        // Dispatch to typed handlers
        if let typed = typedEventHandlers[event.type] {
            for handler in typed.values {
                handler(event)
            }
        }

        // Dispatch to wildcard handlers
        for handler in eventHandlers.values {
            handler(event)
        }
    }

    // MARK: - Tool Registration (Internal)

    /// Registers tool handlers for this session.
    func registerTools(_ tools: [Tool]) {
        toolHandlers.removeAll()
        for tool in tools {
            toolHandlers[tool.name] = tool.handler
        }
    }

    /// Gets a tool handler by name.
    func getToolHandler(_ name: String) -> ToolHandler? {
        return toolHandlers[name]
    }

    // MARK: - Permission Handler (Internal)

    /// Registers a permission handler.
    func registerPermissionHandler(_ handler: @escaping PermissionHandler) {
        permissionHandler = handler
    }

    /// Handles a permission request from the server.
    func handlePermissionRequest(_ requestData: [String: Any]) async throws
        -> PermissionRequestResult
    {
        guard let handler = permissionHandler else {
            return .denied
        }

        let request = PermissionRequest(
            kind: requestData["kind"] as? String ?? "unknown",
            toolCallId: requestData["toolCallId"] as? String,
            rawData: requestData
        )

        do {
            return try await handler(request, sessionId)
        } catch {
            return .denied
        }
    }

    // MARK: - User Input Handler (Internal)

    /// Registers a user input handler.
    func registerUserInputHandler(_ handler: @escaping UserInputHandler) {
        userInputHandler = handler
    }

    /// Handles a user input request from the server.
    func handleUserInputRequest(_ request: UserInputRequest) async throws -> UserInputResponse {
        guard let handler = userInputHandler else {
            throw CopilotError.operationFailed(
                "User input requested but no handler registered")
        }
        return try await handler(request, sessionId)
    }

    // MARK: - Hooks (Internal)

    /// Registers hook handlers.
    func registerHooks(_ hooks: SessionHooks) {
        self.hooks = hooks
    }

    /// Handles a hook invocation from the server.
    func handleHooksInvoke(hookType: String, input: Any?) async throws -> Any? {
        guard let hooks = hooks else {
            return nil
        }

        switch hookType {
        case "preToolUse":
            guard let handler = hooks.onPreToolUse else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(PreToolUseHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        case "postToolUse":
            guard let handler = hooks.onPostToolUse else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(PostToolUseHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        case "userPromptSubmitted":
            guard let handler = hooks.onUserPromptSubmitted else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(
                    UserPromptSubmittedHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        case "sessionStart":
            guard let handler = hooks.onSessionStart else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(SessionStartHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        case "sessionEnd":
            guard let handler = hooks.onSessionEnd else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(SessionEndHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        case "errorOccurred":
            guard let handler = hooks.onErrorOccurred else { return nil }
            if let inputDict = input as? [String: Any] {
                let data = try JSONSerialization.data(withJSONObject: inputDict, options: [])
                let hookInput = try JSONDecoder().decode(ErrorOccurredHookInput.self, from: data)
                let output = try await handler(hookInput, sessionId)
                if let output = output {
                    return encodeToDictionary(output)
                }
            }
            return nil

        default:
            return nil
        }
    }

    // MARK: - Get Metadata

    /// Retrieves metadata for this session.
    ///
    /// - Returns: A dictionary containing the session metadata.
    public func getMetadata() async throws -> [String: Any] {
        return try await rpcClient.sendRequest(
            method: "session.getMetadata",
            params: ["sessionId": sessionId]
        )
    }

    // MARK: - Get Messages

    /// Retrieves all events and messages from this session's history.
    ///
    /// - Returns: An array of all session events.
    public func getMessages() async throws -> [SessionEvent] {
        let response = try await rpcClient.sendRequest(
            method: "session.getMessages",
            params: ["sessionId": sessionId]
        )

        guard let events = response["events"] as? [[String: Any]] else {
            throw CopilotError.invalidResponse("Invalid getMessages response")
        }

        return events.compactMap { SessionEvent.from(dict: $0) }
    }

    // MARK: - Destroy

    /// Destroys this session and releases all associated resources.
    ///
    /// After calling this method, the session can no longer be used. All event
    /// handlers and tool handlers are cleared.
    public func destroy() async throws {
        let _ = try await rpcClient.sendRequest(
            method: "session.destroy",
            params: ["sessionId": sessionId]
        )
        eventHandlers.removeAll()
        typedEventHandlers.removeAll()
        toolHandlers.removeAll()
        permissionHandler = nil
        userInputHandler = nil
        hooks = nil
    }

    // MARK: - Abort

    /// Aborts the currently processing message in this session.
    ///
    /// The session remains valid and can continue to be used for new messages.
    public func abort() async throws {
        let _ = try await rpcClient.sendRequest(
            method: "session.abort",
            params: ["sessionId": sessionId]
        )
    }
}
